stats/stats_linux.go

```diff
@@ -0,0 +1,157 @@
+				if err := s.getLoadAvgInfo(); err != nil {
+					s.loadAvg1Min = "(unknown)"
+					s.loadAvg5Min = "(unknown)"
+					s.loadAvg15Min = "(unknown)"
+					slog.Info("failed to get loadavg", "err", err)
+				}
+
+				if err := s.getMemInfo(); err != nil {
+					s.memTotal = "(unknown)"
+					s.memFree = "(unknown)"
+					s.memAvailable = "(unknown)"
+					slog.Info("failed to get meminfo", "err", err)
+				}
+			case <-ctx.Done():
+			}
+		}
+	}()
+
+	return s
+
+}
+
+func (s *Stats) GetStatsTags() map[string]string {
+	return map[string]string{
+		"_subtrace_linux_num_cpus":              s.numCpus,
+		"_subtrace_linux_loadavg_1_min":         s.loadAvg1Min,
+		"_subtrace_linux_loadavg_5_min":         s.loadAvg5Min,
+		"_subtrace_linux_loadavg_15_min":        s.loadAvg15Min,
+		"_subtrace_linux_meminfo_mem_total":     s.memTotal,
+		"_subtrace_linux_meminfo_mem_free":      s.memFree,
+		"_subtrace_linux_meminfo_mem_available": s.memAvailable,
+	}
+}
+
+func (s *Stats) getNumCpus() error {
+	info, err := os.ReadFile("/proc/cpuinfo")
+	if err != nil {
+		return fmt.Errorf("read /proc/cpuinfo: %w", err)
+	}
+
+	blocks := bytes.Split(info, []byte{'\n', '\n'})
+
+	// Account for newlines at the end of the file
+	if len(blocks[len(blocks)-1]) == 0 {
+		s.numCpus = strconv.Itoa(len(blocks) - 1)
+	} else {
+		s.numCpus = strconv.Itoa(len(blocks))
+	}
+
+	return nil
+}
+
+// See [1] for the format of the /proc/loadavg file.
+//
+// [1] https://man7.org/linux/man-pages/man5/proc_loadavg.5.html
+func (s *Stats) getLoadAvgInfo() error {
+	loadavg, err := os.ReadFile("/proc/loadavg")
+	if err != nil {
+		return fmt.Errorf("read /proc/loadavg: %w", err)
+	}
+
+	values := bytes.Split(loadavg, []byte{' '})
+	if len(values) < 3 {
+		return fmt.Errorf("not enough info in /proc/loadavg, expected at least 3 values but got %d", len(values))
+	}
+
+	for i := 0; i < 3; i++ {
+		value, err := strconv.ParseFloat(string(values[i]), 32)
+		if err != nil {
+			return fmt.Errorf("parse %s as float: %w", string(values[i]), err)
+		}
+
+		switch i {
+		case 0:
+			s.loadAvg1Min = fmt.Sprintf("%.2f", value)
+		case 1:
+			s.loadAvg5Min = fmt.Sprintf("%.2f", value)
+		case 2:
+			s.loadAvg15Min = fmt.Sprintf("%.2f", value)
+		}
+	}
+
+	return nil
+}
+
+// See [1] for the format of the /proc/loadavg file.
+//
+// [1] https://man7.org/linux/man-pages/man5/proc_meminfo.5.html
+func (s *Stats) getMemInfo() error {
+	meminfo, err := os.ReadFile("/proc/meminfo")
+	if err != nil {
+		return fmt.Errorf("read: %w", err)
+	}
+
+	lines := bytes.Split(meminfo, []byte{'\n'})
+	if len(lines) < 3 {
+		return fmt.Errorf("not enough info, expected at least 3 lines but got %d", len(lines))
+	}
+
+	for i := 0; i < 3; i++ {
+		words := bytes.Split(lines[i], []byte{' '})
+		if len(words) < 2 {
+			return fmt.Errorf("split: expected at least 2 words in line but got %d", len(words))
+		}
+
+		value := fmt.Sprintf("%s %s", words[len(words)-2], words[len(words)-1])
+		switch i {
+		case 0:
```

Comment:
this assumes the order of lines is the same. a more robust parser would rely on the string key before the `:` instead of the line number.